dali/core/fast_div_test.cu

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/core/fast_div.h"  // NOLINT
+#include <gtest/gtest.h>
+#include <random>
+#include <iostream>
+#include "dali/test/device_test.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/cuda_event.h"
+
+namespace dali {
+
+template <uint32_t divisor>
+void TestDiv32() {
+  fast_div<uint32_t> fast = divisor;
+  std::cerr << "Testing x / " << divisor << std::endl;
+  for (uint32_t x = 0; x < 0xffffffffu; x++) {
+    ASSERT_EQ(x / divisor, x / fast) << " when dividing " << x << " / " << divisor;
+  }
+}
+
+// This test is disabled because it's ridiculously slow - only run when touching fast_div
+TEST(FastDiv, DISABLED_U32_Host) {
+  // use compile-time constants so the compiler can optimize the reference division
+  TestDiv32<0xFFFFFFFEu>();
+  TestDiv32<3>();
+  TestDiv32<5>();
+  TestDiv32<7>();
+  TestDiv32<1>();
+  TestDiv32<0x800000>();
+  TestDiv32<19>();
+  TestDiv32<42>();
+  TestDiv32<12345678>();
+  TestDiv32<0x82345678>();
+}
+
+template <uint64_t divisor>
+void TestDiv64() {
+  fast_div<uint64_t> fast = divisor;
+  std::cerr << "Testing x / " << divisor << std::endl;
+  for (uint64_t xx = 0; xx < 0xffffffffu; xx++) {
+    uint64_t x = ((xx >> 12) << 44) + (xx & 0xfff);
+    ASSERT_EQ(x / divisor, x / fast) << " when dividing " << x << " / " << divisor;
+  }
+}
+
+// This test is disabled because it's ridiculously slow - only run when touching fast_div
+TEST(FastDiv, DISABLED_U64_Host) {
+  // use compile-time constants so the compiler can optimize the reference division
+  TestDiv64<0xFFFFFFFFFFFFFFFEuL>();
+  TestDiv64<3>();
+  TestDiv64<5>();
+  TestDiv64<7>();
+  TestDiv64<1>();
+  TestDiv64<0x800000>();
+  TestDiv64<0x80000000000000uL>();
+  TestDiv64<19>();
+  TestDiv64<42>();
+  TestDiv64<12345678901234567uL>();
+  TestDiv64<0x82345678DEADBEEFuL>();
+}
+
+
+// This test is disabled because it's ridiculously slow - only run when touching fast_div
+DEVICE_TEST(FastDiv, DISABLED_U32_GPU, dim3(1<<10, 1<<10), (1<<10)) {
+  uint32_t start = (blockIdx.x * blockDim.x + threadIdx.x) << 12;
+  uint32_t divisor = blockIdx.y + 1;
+  fast_div<uint32_t> fast = divisor;
+  uint32_t end = start + (1<<12);
+  if (end == 0)
+    end--;
```

Comment:
Overflow?

Reply:
That's what I'm guarding against. Anyway, this algorithm can't divide (2^N-1) by "uncooperative" numbers, such as 7, 19 and composite numbers with only odd prime factors, one of which is an uncooperative number.